### PR TITLE
small code and docs refactoring

### PR DESCRIPTION
**Pull request by StrikerRUS**

### CODE REVIEW

File: CMakeLists.txt
Comment (StrikerRUS) [inline CMakeLists.txt:134]: Should be set after `-Wall` to take effect.

File: CMakeLists.txt
Comment (StrikerRUS) [inline CMakeLists.txt:270]: Hide

```
/Users/runner/work/1/s/src/objective/binary_objective.hpp:208:14: warning: private field 'deterministic_' is not used [-Wunused-private-field]
  const bool deterministic_;
             ^
```

with AppleClang.

File: include/LightGBM/c_api.h
Comment (StrikerRUS) [inline include/LightGBM/c_api.h:405]: Should be `booster`, not `dataset`.

File: src/io/config.cpp
Comment (StrikerRUS) [inline src/io/config.cpp:345]: To generalize for possible future new enum options.

File: src/io/dataset_loader.cpp
Comment (StrikerRUS) [inline src/io/dataset_loader.cpp:603]: Unused.

File: src/treelearner/linear_tree_learner.cpp
Comment (StrikerRUS) [inline src/treelearner/linear_tree_learner.cpp:214]: Not sure about this. Refer to https://github.com/microsoft/LightGBM/pull/2154#issuecomment-490716927 and https://github.com/microsoft/LightGBM/pull/1739.
@btrotta Is there a possibility to meet equivalent elements in `raw_features`?
cc @guolinke

**Review by jameslamb — COMMENTED** [review]
changes look good to me! But @btrotta should approve before we merge

File: src/treelearner/linear_tree_learner.cpp
Comment (guolinke) [reply src/treelearner/linear_tree_learner.cpp:214]: I think it is not need for stable-sort. I think here is only to get unique feature indices.

Comment (StrikerRUS) [reply src/treelearner/linear_tree_learner.cpp]: @guolinke 
Thanks! Reverted in 4358b4a783074fbd520e4882a79d1092ad9875ff.

**Review by guolinke — APPROVED** [review]

**Pull request merged**
